framework/src/messaging/messaging_recv.c

```diff
@@ -0,0 +1,218 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <tinyara/config.h>
+#include <debug.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <mqueue.h>
+#include <stdlib.h>
+#include <signal.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <messaging/messaging.h>
+#include "messaging_internal.h"
+
+/****************************************************************************
+ * private functions
+ ****************************************************************************/
+static int set_notification(messaging_recv_info_t *data)
+{
+	int ret;
+	struct sigevent notification;
+
+	notification.sigev_notify = SIGEV_SIGNAL;
+	notification.sigev_signo = SIGMSG_UNICAST;
+	notification.sigev_value.sival_ptr = data;
+
+	ret = mq_notify(data->mqdes, &notification);
+	if (ret < 0) {
+		msgdbg("message recv fail : set notification fail. %d\n", errno);
+		MSG_FREE(data);
+		mq_close(data->mqdes);
+		return ERROR;
+	}
+	return OK;
+}
+
+static void wrapper_callback(int signo, siginfo_t *data)
+{
+	int ret;
+	messaging_recv_info_t *info;
+	char msg[CONFIG_MESSAGING_UNICAST_MSGSIZE];
+	struct mq_attr attr;
+
+	info = (messaging_recv_info_t *)data->si_value.sival_ptr;
+	if (info == NULL) {
+		msgdbg("message recv fail : wrong param for wrapper callback.\n");
+		return;
+	}
+
+	while (1) {
+		ret = mq_receive(info->mqdes, (char *)&msg, sizeof(messaging_recv_info_t), 0);
+		if (ret < 0) {
+			msgdbg("message recv fail : mq_receive fail. %d\n", errno);
+			mq_close(info->mqdes);
+			MSG_FREE(data);
+			return;
+		}
+		/* Call user callback */
+		(*info->user_cb)(NULL, &msg);
+
+		ret = mq_getattr(info->mqdes, &attr);
+		if (ret < 0) {
+			msgdbg("message recv fail : mq_getattr fail.\n");
+			mq_close(info->mqdes);
+			MSG_FREE(data);
+			return;
+		} else if (attr.mq_curmsgs == 0) {
+			/* All requests are handled. Register notification again. */
+			set_notification(info);
+			break;
+		}
+	}
+}
+
+static int set_notify_signal(void)
+{
+	int ret;
+	struct sigaction act;
+	act.sa_sigaction = (_sa_sigaction_t)wrapper_callback;
+	act.sa_flags = 0;
+	(void)sigemptyset(&act.sa_mask);
+
+	ret = sigaddset(&act.sa_mask, SIGMSG_UNICAST);
+	if (ret < 0) {
+		msgdbg("message recv fail : sigaddset error.\n");
+		return ERROR;
+	}
+
+	ret = sigaction(SIGMSG_UNICAST, &act, NULL);
+	if (ret == (int)SIG_ERR) {
+		msgdbg("message recv fail : sigaction error.\n");
+		return ERROR;
+	}
+	return OK;
+}
+
```

Comment:
Same as above, as well as input/output parameters details, please do it for all the functions.

Reply:
Done.